ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,884 @@
+        }
+        writtenLastTime = srcValue.getSize();
+      } else if (!isActive) {
+        logger.info("Stopping copying because shutdown is in progress");
+        copiedAll = false;
+        break;
+      } else {
+        // this is the extra segment, so it is ok to run out of space.
+        logger.info("There is no more capacity in the destination log. Total capacity is {}. Used capacity is {}. "
+            + "Segment that was being copied is {}", totalCapacity, usedCapacity, logSegmentToCopy.getName());
+        copiedAll = false;
+        break;
+      }
+    }
+    logSegmentToCopy.closeView();
+    return copiedAll;
+  }
+
+  // commit() helpers
+
+  /**
+   * Gets the names of the target log segments that were generated (the swap spaces). Note that this function returns
+   * those log segments that already exist and is not useful to determine what should exist and it only returns the ones
+   * that are still in their temporary uncommitted state.
+   * @return the list of target log segments that exist on the disk.
+   */
+  private List<String> getTargetLogSegmentNames() {
+    List<String> filenames = Arrays.asList(dataDir.list(TEMP_LOG_SEGMENTS_FILTER));
+    List<String> names = new ArrayList<>(filenames.size());
+    for (String filename : filenames) {
+      String transformed = filename.substring(0, filename.length() - TEMP_LOG_SEGMENT_NAME_SUFFIX.length());
+      names.add(LogSegmentNameHelper.nameFromFilename(transformed));
+    }
+    Collections.sort(names, LogSegmentNameHelper.COMPARATOR);
+    return names;
+  }
+
+  /**
+   * Renames the temporary log segment files to remove the temporary file name suffix and effectively commits them on
+   * disk.
+   * @param logSegmentNames the names of the log segments whose backing files need to be renamed.
+   * @throws StoreException if the rename fails.
+   */
+  private void renameLogSegments(List<String> logSegmentNames) throws StoreException {
+    for (String segmentName : logSegmentNames) {
+      String filename = LogSegmentNameHelper.nameToFilename(segmentName);
+      File from = new File(dataDir, filename + TEMP_LOG_SEGMENT_NAME_SUFFIX);
+      File to = new File(dataDir, filename);
+      if (!from.renameTo(to)) {
+        // TODO: change error code
+        throw new StoreException("Failed to rename " + from + " to " + to, StoreErrorCodes.Unknown_Error);
+      }
+    }
+  }
+
+  /**
+   * Commits the log segments with names {@code logSegmentNames}. Committing involves adding them to the application log
+   * instance.
+   * @param logSegmentNames the names of the log segments to commit.
+   * @param recovering {@code true} if this function was called in the context of recovery. {@code false} otherwise.
+   * @throws IOException if there were any I/O errors creating a log segment
+   */
+  private void commitLogSegments(List<String> logSegmentNames, boolean recovering) throws IOException {
```

Comment:
The persistent commit has already been made in `renameLogSegments()`. I'm not opposed to changing the name of this function however. What do you think is a good name?